### PR TITLE
 cpu/stm32: add rtc_mem

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

WIP of rtc_mem implementation for stm32, just to see if the API works fine for stm32 BOARDs as well.

### Testing procedure

Both tests/periph_rtc and tests/periph_rtt have been adapted to make use of the feature and to ensure it does not interfere with normal RTC/RTT alarms.

Need to see if those REGS do not exist on some families....

### Issues/PRs references

Depends on https://github.com/RIOT-OS/RIOT/pull/16758

### CODE REVIEW

File: cpu/stm32/Makefile.features [hunk 1]
```diff
@@ -17,6 +17,13 @@ ifneq (,$(filter $(CPU_FAM),f0 f1 f3 g0 g4 l0 l1 l4 l5 wb wl))
   FEATURES_PROVIDED += periph_flashpage_pagewise
 endif
 
+ifneq (,$(filter $(CPU_FAM),f0 f3 g0 g4 l0 l1 l4 l5 wb wl))
+  CPUT_MODELS_WITHOUT_RTC_BKPR += stm32f030% stm32f070%
+  ifeq (,$(filter $(CPU_MODELS_WITHOUT_HWRNG),$(CPU_MODEL)))
```
Comment (benpicco) [inline cpu/stm32/Makefile.features:22]:
```suggestion
  ifeq (,$(filter $(CPUT_MODELS_WITHOUT_RTC_BKPR),$(CPU_MODEL)))
```

File: cpu/stm32/periph/rtc_all.c [hunk 2]
```diff
@@ -209,7 +209,7 @@ static int bcd2val(uint32_t val, int shift, uint32_t mask)
     return (((tmp >> 4) * 10) + (tmp & 0x0f));
 }
 
-static inline void rtc_unlock(void)
+extern inline void rtc_unlock(void)
```
Comment (benpicco) [inline cpu/stm32/periph/rtc_all.c:212]: `extern inline void` sounds like an oxymoron to me. What does it do?

File: cpu/stm32/Makefile.features (hunk 1 above)
Comment (fjmolinas) [reply cpu/stm32/Makefile.features]: copy pasta thanks

File: cpu/stm32/periph/rtc_all.c (hunk 2 above)
Comment (fjmolinas) [reply cpu/stm32/periph/rtc_all.c:212]: Another copy pasta, did not think it through...

Comment (fjmolinas) [reply cpu/stm32/periph/rtc_all.c:212]: Not sure why it was inlined should I keep that and move the function to stm32 header?

File: cpu/stm32/periph/rtc_all.c [hunk 3]
```diff
@@ -221,7 +221,7 @@ static inline void rtc_unlock(void)
     while (!(RTC_REG_ISR & RTC_ISR_INITF)) {}
 }
 
-static inline void rtc_lock(void)
+extern void rtc_lock(void)
```
Comment (benpicco) [inline cpu/stm32/periph/rtc_all.c:224]: I still don't get what the `extern` is doing here

Comment (fjmolinas) [reply cpu/stm32/periph/rtc_all.c:224]: You say because the `extern` is implicit? Ok will remove.

File: cpu/stm32/periph/rtc_all.c (hunk 2 above)
Comment (benpicco) [reply cpu/stm32/periph/rtc_all.c]: The function does quite a few things, I think there is no reason to force it to be inlined.
We should even save some ROM this way.

File: cpu/stm32/periph/rtc_mem.c [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+extern void rtc_lock(void);
+extern void rtc_unlock(void);
+
+size_t rtc_mem_size(void)
+{
+    return RTC_MEM_SIZE * __SIZEOF_POINTER__;
+}
+
+void rtc_mem_write(unsigned offset, const void *data, size_t len)
+{
```
Comment (benpicco) [inline cpu/stm32/periph/rtc_mem.c:102]: > 'data' is of type 'const void *'. When using void pointers in calculations, the behaviour is undefined.

Looks like you need something like

```suggestion
{
    const uint8_t *in = data;
```

**Comment by benpicco** [comment]
ah Kconfig needs an update too

File: cpu/stm32/periph/rtc_mem.c [hunk 5]
```diff
@@ -104,7 +104,8 @@ void rtc_mem_write(unsigned offset, const void *data, size_t len)
         return;
     }
 
-    /* get the correct pointer */
+    uint8_t *in = (uint8_t *)data;
```
Comment (benpicco) [inline cpu/stm32/periph/rtc_mem.c:107]: better keep that
```suggestion
    const uint8_t *in = data;
```

Comment (fjmolinas) [reply cpu/stm32/periph/rtc_mem.c:107]: It will lead to a cast in the memcpy call, is that better?

Comment (benpicco) [reply cpu/stm32/periph/rtc_mem.c:107]: but the second argument of `memcpy` is `const`

Comment (fjmolinas) [reply cpu/stm32/periph/rtc_mem.c:107]: duh

Comment (fjmolinas) [reply cpu/stm32/periph/rtc_mem.c:107]: I'll addressm Kconfig now matches, do you mind if I squash once the const is added?

Comment (fjmolinas) [reply cpu/stm32/periph/rtc_mem.c]: done

**Review by benpicco — APPROVED** [review]
Tested on `nucleo-l031k6` and works fine!

**Pull request merged**